### PR TITLE
feat(x86_64): read memory map from bootloader

### PR DESCRIPTION
**Pull request by hawkw**
depends on #27

the x64 kernel now reads the memory map from the bootloader and converts
it into our HAL types. i also implemented some unimplemented functions
in order to make this work.

Screen Shot 2019-12-13 at 11 07 54 AM

### CODE REVIEW

File: hal-x86_64/src/lib.rs [hunk 1]
```diff
@@ -18,9 +18,15 @@ impl Architecture for X64 {
 
 impl fmt::Debug for PAddr {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_tuple("PAddr")
-            .field(&format_args!("{:#x}", self.0))
-            .finish()
+        if let Some(width) = f.width() {
+            f.debug_tuple("PAddr")
+                .field(&format_args!("{:#0width$x}", self.0, width = width))
```
Comment (iximeow) [inline hal-x86_64/src/lib.rs]: woah til padding length can be variable

File: hal-core/src/mem/mod.rs [hunk 2]
```diff
@@ -112,3 +127,19 @@ impl RegionKind {
...
 }
+
+impl fmt::Debug for RegionKind {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            &RegionKind::FREE => f.pad("FREE"),
+            &RegionKind::USED => f.pad("USED"),
+            &RegionKind::BOOT => f.pad("BOOT"),
+            &RegionKind::BOOT_RECLAIMABLE => f.pad("BOOT_RECLAIMABLE"),
+            &RegionKind::BAD => f.pad("BAD T_T"),
```
Comment (iximeow) [inline hal-core/src/mem/mod.rs]: T_T

File: hal-core/src/mem/mod.rs [hunk 3]
```diff
@@ -112,3 +127,19 @@ impl RegionKind {
     /// Memory used for a page table.
     pub const PAGE_TABLE: Self = Self(KindInner::PageTable);
 }
+
+impl fmt::Debug for RegionKind {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
```
Comment (iximeow) [inline hal-core/src/mem/mod.rs]: do you know how mmio shows up here? is this not a concern of the memory map at this point, pending further bootstrap, or is  it `USED`?

File: hal-x86_64/src/lib.rs (hunk 1 above)
Comment (hawkw) [reply hal-x86_64/src/lib.rs]: fmt specs are surprisingly magical

File: hal-core/src/mem/mod.rs (hunk 3 above)
Comment (hawkw) [reply hal-core/src/mem/mod.rs]: i _think_ they will show up as `Reserved` in the bootloader crate's memory map: https://github.com/rust-osdev/bootloader/blob/90f5b8910d146d6d489b70a6341d778253663cfa/src/bootinfo/memory_map.rs#L174-L175
but it isn't particularly clear from the comment. would have to actually look at the source to figure that out.

i think we'll want to get most mmio ranges from ACPI or BDA (does anyhting use that besides COM ports?) though.

**Pull request merged**